provisioning_test.go

```diff
@@ -576,3 +510,178 @@ func TestProvisionWithInvalidEkCert(t *testing.T) {
+				Attrs: tpm2.AttrFixedTPM | tpm2.AttrFixedParent | tpm2.AttrSensitiveDataOrigin | tpm2.AttrUserWithAuth | tpm2.AttrNoDA |
+					tpm2.AttrRestricted | tpm2.AttrDecrypt,
+				Params: &tpm2.PublicParamsU{
+					RSADetail: &tpm2.RSAParams{
+						Symmetric: tpm2.SymDefObject{
+							Algorithm: tpm2.SymObjectAlgorithmAES,
+							KeyBits:   &tpm2.SymKeyBitsU{Sym: 128},
+							Mode:      &tpm2.SymModeU{Sym: tpm2.SymModeCFB}},
+						Scheme:   tpm2.RSAScheme{Scheme: tpm2.RSASchemeNull},
+						KeyBits:  2048,
+						Exponent: 0}}}
+
+			if err := tpm.EnsureProvisionedWithCustomSRK(data.mode, nil, &template); err != nil {
+				t.Errorf("EnsureProvisionedWithCustomSRK failed: %v", err)
+			}
+
+			validatePrimaryKeyAgainstTemplate(t, tpm.TPMContext, tpm2.HandleOwner, tcg.SRKHandle, &template)
+
+			key, _, _, _, _, err := tpm.CreatePrimary(tpm.OwnerHandleContext(), nil, &template, nil, nil, nil)
+			if err != nil {
+				t.Fatalf("CreatePrimary failed: %v", err)
+			}
+			defer flushContext(t, tpm, key)
+
+			srk, err := tpm.CreateResourceContextFromTPM(tcg.SRKHandle)
+			if err != nil {
+				t.Fatalf("CreateResourceContextFromTPM failed: %v", err)
+			}
+
+			if !bytes.Equal(srk.Name(), key.Name()) {
+				t.Errorf("Unexpected SRK ID")
+			}
+
+			nv, err := tpm.CreateResourceContextFromTPM(0x01810001)
+			if err != nil {
+				t.Fatalf("CreateResourceContextFromTPM failed: %v", err)
+			}
+
+			nvPub, _, err := tpm.NVReadPublic(nv)
+			if err != nil {
+				t.Fatalf("NVReadPublic failed: %v", err)
+			}
+
+			tmplB, err := tpm.NVRead(tpm.OwnerHandleContext(), nv, nvPub.Size, 0, nil)
+			if err != nil {
+				t.Errorf("NVRead failed: %v", err)
+			}
+
+			expected, _ := mu.MarshalToBytes(&template)
+			if !bytes.Equal(tmplB, expected) {
+				t.Errorf("Unexpected template")
+			}
+		})
+	}
+}
+
+func TestProvisionWithInvalidCustomSRKTemplate(t *testing.T) {
+	tpm, _ := openTPMSimulatorForTesting(t)
+	defer func() {
+		clearTPMWithPlatformAuth(t, tpm)
+		closeTPM(t, tpm)
+	}()
+
+	clearTPMWithPlatformAuth(t, tpm)
+
+	template := tpm2.Public{
+		Type:    tpm2.ObjectTypeRSA,
+		NameAlg: tpm2.HashAlgorithmSHA256,
+		Attrs: tpm2.AttrFixedTPM | tpm2.AttrFixedParent | tpm2.AttrSensitiveDataOrigin | tpm2.AttrUserWithAuth | tpm2.AttrNoDA |
+			tpm2.AttrRestricted | tpm2.AttrSign,
+		Params: &tpm2.PublicParamsU{
+			RSADetail: &tpm2.RSAParams{
+				Symmetric: tpm2.SymDefObject{
+					Algorithm: tpm2.SymObjectAlgorithmAES,
+					KeyBits:   &tpm2.SymKeyBitsU{Sym: 128},
+					Mode:      &tpm2.SymModeU{Sym: tpm2.SymModeCFB}},
+				Scheme:   tpm2.RSAScheme{Scheme: tpm2.RSASchemeNull},
+				KeyBits:  2048,
+				Exponent: 0}}}
+	err := tpm.EnsureProvisionedWithCustomSRK(ProvisionModeFull, nil, &template)
+	if err == nil {
+		t.Fatalf("EnsureProvisionedWithCustomSRK should have failed")
+	}
+
+	if err.Error() != "supplied SRK template is not valid for a parent key" {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+}
+
+func TestRemoveStoredCustomSRKTemplate(t *testing.T) {
```

Comment:
shouldn't this also be called something like TestProvisionDefaultRemoves... ?

Reply:
I've renamed this, but thinking about it a bit more - I'm not sure that the behaviour implemented here is actually correct. I think that EnsureProvisioned should use the custom template if EnsureProvisionedWithCustomSRK was used previously and the TPM isn't being cleared. This would make it possible to inject a custom template into an image, execute EnsureProvisionedWithCustomSRK on first boot and then be able to just use the existing simpler API after that if it's required, without having do do things like "if hasCustomTemplate { EnsureProvisionedWithCustomSRK() } else { EnsureProvisioned() }"

Reply:
yes, your new suggested behavior seems reasonable as well

Reply:
Ok, I've updated this to change the behaviour now